File: math/dm13/dm.py
#!/usr/bin/python3
from math import sqrt;

def norm(l):
    i = len(l) - 1
    while i >= 0 and l[i] == 0:
        i -= 1
    if i < 0:
        return [0]
    else:
        return l[:i+1]

def multXp(a, p, coeff):
    l = [0]*p + a
    for i in range(p, len(l)):
        l[i] *= coeff
    return l

def addP(a, b):
    n = max(len(a),len(b))
    c = [0]*n
    for i in range(n):
        if i < len(a):
            c[i] = a[i]
        if i < len(b):
            c[i] += b[i]
    return c

def divEuclid(a, b):
    q = [0]*(len(a) - len(b) + 1)
    r = a
    while len(r) >= len(b):
        p = len(r) - len(b)
        c = r[-1]/b[-1]
        q[p] = c
        m = multXp(b, p, -c)
        r = addP(r, m)
        r = norm(r)
    return (q,r)

def compXp(a, p):
    l = [0]*((len(a) - 1) * p + 1)
    for i in range(len(a)):
        l[i*p] = a[i]
    return l

def polyCyclo(l):
    if len(l) == 0:
        return [-1, 1]
    p = l[0][0]
    ls = l[1:]
    for i in range(len(l)):
        p *= l[i][0] ** (l[i][1] - 1)
    poly = polyCyclo(ls)
    return divEuclid(compXp(poly, p), poly)[0]

def decomp(n):
    if n == 1:
        return []
    def joindecomp(d1, d2):
        d = d1
        for i in range(len(d2)):
            p = -1
            for j in range(len(d)):
                if d[j][0] == d2[i][0]:
                    p = j
                    break
            if p == -1:
                d.append(d2[i])
            else:
                d[p][1] += d2[i][1]
        return d

    s = int(sqrt(n))
    for d in range(2, s+1):
        if n%d == 0:
            return joindecomp(decomp(d), decomp(n//d))
    # n prime
    return [[n,1]]

def polyCyclo2(n):
    return polyCyclo(decomp(n))

def merge(l, out):
    for a in out:
        if not a in l:
            l.append(a)

def quicksort(l):
    if len(l) <= 1:
        return l
    a = l[0]
    l1 = [x for x in l if x < a]
    m  = [x for x in l if x == a]
    l2 = [x for x in l if x > a]
    return quicksort(l1) + m + quicksort(l2)

if __name__ == "__main__":
    coeffs = []
    test = True
    for n in range(2,100):
        cfs = polyCyclo2(n)
        if test:
            for a in cfs:
                if not a in [-1, 0, 1]:
                    test = False
                    print("First other :", n)
                    break
        merge(coeffs, cfs)
    print(quicksort(coeffs))

File: math/dm13/dm.hs
type Corps = Double
type Poly = [Corps]

zeros :: Int -> [Corps]
zeros n = fmap (\x -> 0) [1..n]

deg :: Poly -> Int
deg p = length p

dom :: Poly -> Corps
dom p = p!!(deg p - 1)

norm :: Poly -> Poly
norm [l] = [l]
norm l   = normalize 0 l
    where normalize :: Int -> Poly -> Poly
          normalize _ []     = []
          normalize n (0:ls) = normalize (n+1) ls
          normalize n (l:ls) = zeros n ++ [l] ++ normalize 0 ls

multXp :: Poly -> Int -> Corps -> Poly
multXp [] _ _ = []
multXp pl p l = zeros p ++ fmap (*l) pl

addP :: Poly -> Poly -> Poly
addP []     b      = b
addP a      []     = a
addP (a:as) (b:bs) = (a+b) : addP as bs

divEuclid :: Poly -> Poly -> (Poly,Poly)
divEuclid a b = mdiv [] a b
    where mdiv :: Poly -> Poly -> Poly -> (Poly,Poly)
          mdiv q a b = if deg a < deg b then (q, a)
                       else mdiv q2 r b
              where d  = deg a - deg b
                    c  = dom a / dom b
                    r  = hd $ addP a $ multXp b d (-1*c)
                    q2 = c:q
                    hd (a:as) = if length as == 0 then [] else a:hd as

compXp :: Int -> Poly -> Poly
compXp n p = norm $ (p!!0):[value i | i <- [1..(n*deg p - 1)]]
    where value i = if i `mod` n == 0 then (p!!(div i n)) else 0

first :: (a, b) -> a
first (x, _) = x

pow :: Int -> Int -> Int
pow _ 0 = 1
pow x n = x * pow x (n-1)

polyCyclo :: [(Int,Int)] -> Poly
polyCyclo cfs = compXp cf $ subPolyCyclo $ fmap first cfs
    where cf = foldl (*) 1 [pow pl (al - 1) | (pl,al) <- cfs]
          subPolyCyclo :: [Int] -> Poly
          subPolyCyclo []     = [-1, 1]
          subPolyCyclo (p:ps) = first $ divEuclid (compXp p pl) pl
              where pl = subPolyCyclo ps

nsqrt :: Int -> Int
nsqrt n = truncate $ sqrt $ fromRational $ toRational n

sortFus :: [a] -> (a -> a -> Ordering) -> (a -> a -> a) -> [a]
sortFus []     _ _ = []
sortFus (l:ls) c f = sortFus l1 c f ++ [foldl f l m] ++ sortFus l2 c f
    where l1 = [x | x <- ls, x `c` l == LT]
          m  = [x | x <- ls, x `c` l == EQ]
          l2 = [x | x <- ls, x `c` l == GT]

decomp :: Int -> [(Int,Int)]
decomp n = sortFus (subDecomp n) cmp fus
    where subDecomp 1 = []
          subDecomp n = if length dvs == 0 then [(n,1)]
                        else let (d,d2) = ((dvs!!0), n `div` d) in decomp d ++ decomp d2
              where dvs = [d | d <- [2..nsqrt n], n `mod` d == 0]
          cmp (a1,_) (a2,_)   = a1 `compare` a2
          fus (a1,b1) (a2,b2) = (a1, b1+b2)

polyCyclo2 :: Int -> Poly
polyCyclo2 n = polyCyclo $ decomp n

main :: IO()
main = putStrLn $ show $ polyCyclo2 60

